docs/Lifecycling.md

```diff
@@ -0,0 +1,148 @@
+
+Thus, exactly at time `t_1`, the condition on the outer `when` is met and we acquire `c2`
+
+```Haskell
+Scale (Observe "USD_LIBOR_3M") (When (t ≥ t_2) (One “USD”))
+```
+
+as replacement for the initial contract.
+
+As soon as we acquire `c2`, because of the `scale` node, we acquire `c3`
+
+```Haskell
+USD_LIBOR_3M(t_1) * (When (t ≥ t_2) (One “USD”)
+```
+
+as a replacement for `c2`.
+
+Finally, once `t_2` has arrived, the condition on the remaining `when` is met and we acquire `c4`
+
+```Haskell
+USD_LIBOR_3M(t_1)) * (One “USD”)
+```
+
+as a replacement for `c3`.
+
+This contract does not include additional dynamics and will not evolve further. In a financial setting, this is when a payment is made to the contract owner for the amount due:
+
+- USD_LIBOR_3M(t_1) USD is paid to the contract holder
+
+- `c4` is replaced with `Zero`, the contract indicating the absence of a claim.
+
+## Lifecycle functionality
+
+In order to replicate this contract evolution process for a generic contract, we need to address two problems:
+
+- given a contract, we need a way to figure out when a relevant event has happened (specifically, the next relevant event date)
+
+- then, we need to be able to evolve the contract according to the rules defined by our contract primitives
+
+The lifecycling functionality of the library addresses the latter problem.
+
+### Types of event dates
+
+Assuming that once we acquire a `One a`, this is immediately paid to the contract holder, We can classify event dates (or time instants) as
+
+- expiry date: when the contract evolves to `Zero` or is equivalent (in the valuation semantics sense) to the `Zero` contract
+
+- payment date: when a `One a` is acquired (usually multiplied by some constant amount)
+
+- exercise date: when the contract holder makes an election (`Or` or `Anytime` nodes)
+
+- fixing date: when the value of an observable is evaluated (usually after a condition on a `When` block becomes true)
+
+In the example above, `t_1` would be a fixing date and `t_2` a payment date, as well as the expiry date.
+
+### Determining event dates
+
+Once we have acquired a contract `c` at a time `t_0`, the next event date is due to either
+
+- an election being made on an `Anytime` node by the contract holder
+
+- a condition within a `When` or `Until` node becoming `True`
+
+- the contract becoming equivalent to the `Zero` contract (e.g. if the probability of the condition within a `When` node becoming `True` is zero)
+
+Event dates related to `When` and `Until` conditions are modelled as stopping times of stochastic processes: these processes need to be continuously observed in order to determine when the event occurs. An exception to this are cases where the next event date is a deterministic time, for instance when using the condition `t ≥ t_1`: in this case, even before `t_1` we know that `t_1` will be an event date.
```

Comment:
There's an interesting reference in "An adventure in financial engineering", to Elliot and Hudak. In section 2.1, they discuss a data type for `Time` that allows them to calculate fixing times from more complex (deterministic) expressions. I haven't had time to look at it in depth, but I thought it would be useful to us.

Reply:
Thanks, I will have a look at this paper